### DISCUSSION TITLE
How to correctly fit and evaluate time series predictor on one very long series?

### DISCUSSION THREAD

**Discussion by evindor** [autogluon/autogluon#4332] [Q&A] [answered]
I have one very long time series, which I would like to chop into much smaller series both for testing and training. I am a bit confused about how this works, especially for testing. Normally I would do this manually, but with AutoGluon it looks like the library does it for us?  Or do I have to manually split my data into multiple series with different `item_id`s?

For example, if I have 1 million data points, and I wish to do a 20-timestep long forecast based on 180-timestep historical data. My understanding that for training AutoGluon uses a sliding window using all available data, so I don't have to do anything and just provide the whole series with a single `item_id`?

But then when using `evaluate`, documentation states that "The last prediction_length time steps of each time series in data will be held out for prediction and forecast accuracy will be calculated on these time steps." So do I have to manually split my long test dataframe into `1_000_000/200 = 5000` short series each with different `item_id` to evaluate the predictor on the test set?

Many thanks for your help.

**Comment by shchur** [comment, marked as answer]
Hi @evindor, AutoGluon currently uses the expanding window splitter, so if you just provide the entire series with one `item_id` it will use all the available historic data, not just the last 180 observations. If that's okay for you, you can use the arguments `num_val_windows`, `val_step_size` and `refit_every_n_windows` to [`TimeSeriesPredictor.fit()`](https://auto.gluon.ai/stable/api/autogluon.timeseries.TimeSeriesPredictor.fit.html) to configure this validation behavior. 

If you want to ensure that only the last 180 time steps are used as input in each window, you will need to write some custom code similar to the following:
```python
steps_per_window = 200
num_observations = df.num_timesteps_per_item().max()
num_windows = num_observations / steps_per_window

all_dfs = []
for window_end_idx in range(steps_per_window, num_observations, steps_per_window):
    df_for_window = df.slice_by_timestep(
        start_index=window_end_idx - steps_per_window, 
        end_index=window_end_idx,
    )
    new_item_id = df_for_window.index.levels[0].astype(str) + f"_index_{window_end_idx}"
    df_for_window.index = df_for_window.index.set_levels(levels=new_item_id, level=0)
    all_dfs.append(df_for_window)
combined_df = pd.concat(all_dfs)
```

**Comment by evindor** [comment]
Thank you for detailed explanation!
